src/renate/memory/buffer.py

```diff
@@ -161,34 +84,50 @@ def __getitem__(self, idx: int) -> Tuple[NestedTensors, DataDict]:
-        _insert_data_point(self._data_points, idx, data)
-        _insert_data_point(self.metadata, idx, metadata)
-
-    def _append(self, data: NestedTensors, metadata: DataDict) -> None:
-        """Appends a data point to the internal storage."""
+    def update(self, dataset: Dataset, metadata: Optional[Dict] = None) -> None:
+        """Updates the buffer with a new dataset."""
+        metadata = metadata or {}
+        self._check_metadata_internal_consistency(metadata, expected_length=len(dataset))
         if not len(self):
-            self._data_points = _make_storage(data, self._max_size)
-            self.metadata = _make_storage(metadata, self._max_size)
-        self[self._size] = data, metadata
-        self._size += 1
+            self._data_point_prototype = copy.deepcopy(dataset[0])
+            self._add_metadata_like(metadata)
+        else:
+            self._check_metadata_compatibility(metadata)
+        assignments = self._update(dataset, metadata)
+        self._check_assignments_are_contiguous(assignments)
+
+        # Perform assignments.
+        d = len(self._datasets)
+        self._datasets.append(dataset)
+        for buffer_idx, dataset_idx in assignments.items():
+            self._indices[buffer_idx] = (d, dataset_idx)
+            for key in self._metadata.keys():
+                self._metadata[key][buffer_idx] = metadata[key][dataset_idx]
 
-    def update(self, dataset: Dataset, metadata: Optional[DataDict] = None) -> None:
-        """Updates the buffer with a new dataset.
+    def _update(self, dataset: Dataset, metadata: Dict) -> Dict[int, int]:
+        """Returns the updates.
+
+        This method is used to implement different buffering methods. Given a dataset, it returns a
+        dictionary of `assignments` indicating the updates to be performed to the buffer.
 
         Args:
-            dataset: A dataset containing a new chunk of data.
-            metadata: A dictionary mapping to tensors, which are assumed to
-                have size n in dimension 0, where `n = len(dataset)`.
+            dataset: The dataset used to update the buffer.
+
+        Returns:
+            assignments: A dictionary mapping int to int. The presence of the pair `(i, j)` means
+                that the `i`-th buffer slot will be filled with the `j`-th element of `dataset`.
         """
-        metadata = metadata or {}
-        return self._update(dataset, metadata)
+        raise NotImplementedError()
+
+    def get_metadata(self, key: str) -> Dict[str, torch.Tensor]:
+        return self._metadata[key][: len(self)]
 
-    @abstractmethod
-    def _update(self, dataset: Dataset, metadata: DataDict) -> None:
-        pass
+    def set_metadata(self, key: str, values: torch.Tensor) -> None:
+        if values.size(0) != len(self):
+            raise ValueError()
+        if key not in self._metadata:
+            self._add_metadata_like({key: values})
+        self._metadata[key][: len(self)] = values.cpu()
```

Comment:
class docstring claims that devices are not changed

Reply:
I guess that passage of the doc string is obsolete. Technically the buffer is now agnostic to the device placement of the data (not the metadata). Metadata is always kept on CPU, so it makes sense to ensure this here so that people can input metadata values (e.g., logits) that reside on GPU.